File: server/routes/api/tasks.js
const express = require('express');
const router = express.Router();
const Task = require('../../models/Task');
const { protect } = require('../../middleware/auth');
const { Op } = require('sequelize');

// @route   GET api/tasks
// @desc    Get all user tasks
// @access  Private
router.get('/', protect, async (req, res) => {
  try {
    const tasks = await Task.findAll({
      where: { userId: req.user.id },
      order: [['dueDate', 'ASC']]
    });
    
    res.json({
      success: true,
      count: tasks.length,
      data: tasks,
    });
  } catch (err) {
    console.error(err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Server error',
    });
  }
});

// @route   POST api/tasks
// @desc    Create a task
// @access  Private
router.post('/', protect, async (req, res) => {
  const {
    title,
    description,
    priority,
    status,
    dueDate,
    reminderTime,
    tags,
    isRecurring,
    recurringPattern,
  } = req.body;

  try {
    const task = await Task.create({
      userId: req.user.id,
      title,
      description,
      priority,
      status,
      dueDate,
      reminderTime,
      tags,
      isRecurring,
      recurringPattern,
    });

    res.status(201).json({
      success: true,
      data: task,
    });
  } catch (err) {
    console.error(err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Server error',
    });
  }
});

// @route   GET api/tasks/:id
// @desc    Get task by ID
// @access  Private
router.get('/:id', protect, async (req, res) => {
  try {
    const task = await Task.findByPk(req.params.id);

    if (!task) {
      return res.status(404).json({
        success: false,
        error: 'Task not found',
      });
    }

    // Make sure user owns task
    if (task.userId !== req.user.id) {
      return res.status(401).json({
        success: false,
        error: 'Not authorized',
      });
    }

    res.json({
      success: true,
      data: task,
    });
  } catch (err) {
    console.error(err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Server error',
    });
  }
});

// @route   PUT api/tasks/:id
// @desc    Update task
// @access  Private
router.put('/:id', protect, async (req, res) => {
  const {
    title,
    description,
    priority,
    status,
    dueDate,
    reminderTime,
    tags,
    isRecurring,
    recurringPattern,
    completedAt,
  } = req.body;

  try {
    let task = await Task.findByPk(req.params.id);

    if (!task) {
      return res.status(404).json({
        success: false,
        error: 'Task not found',
      });
    }

    // Make sure user owns task
    if (task.userId !== req.user.id) {
      return res.status(401).json({
        success: false,
        error: 'Not authorized',
      });
    }

    // Update task fields
    if (title !== undefined) task.title = title;
    if (description !== undefined) task.description = description;
    if (priority !== undefined) task.priority = priority;
    if (status !== undefined) task.status = status;
    if (dueDate !== undefined) task.dueDate = dueDate;
    if (reminderTime !== undefined) task.reminderTime = reminderTime;
    if (tags !== undefined) task.tags = tags;
    if (isRecurring !== undefined) task.isRecurring = isRecurring;
    if (recurringPattern !== undefined) task.recurringPattern = recurringPattern;
    if (completedAt !== undefined) task.completedAt = completedAt;

    await task.save();

    res.json({
      success: true,
      data: task,
    });
  } catch (err) {
    console.error(err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Server error',
    });
  }
});

// @route   DELETE api/tasks/:id
// @desc    Delete task
// @access  Private
router.delete('/:id', protect, async (req, res) => {
  try {
    const task = await Task.findByPk(req.params.id);

    if (!task) {
      return res.status(404).json({
        success: false,
        error: 'Task not found',
      });
    }

    // Make sure user owns task
    if (task.userId !== req.user.id) {
      return res.status(401).json({
        success: false,
        error: 'Not authorized',
      });
    }

    await task.destroy();

    res.json({
      success: true,
      data: {},
    });
  } catch (err) {
    console.error(err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Server error',
    });
  }
});

module.exports = router; 

File: flask-backend/routes/tasks.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from werkzeug.exceptions import BadRequest, NotFound, Forbidden
from sqlalchemy import desc
from models import Task, db
from datetime import datetime

tasks_bp = Blueprint('tasks', __name__)

@tasks_bp.route('/', methods=['GET'])
@jwt_required()
def get_tasks():
    """Get all tasks for the authenticated user"""
    try:
        user_id = get_jwt_identity()
        
        # Get query parameters
        completed = request.args.get('completed')
        priority = request.args.get('priority')
        category = request.args.get('category')
        
        # Build query
        query = Task.query.filter_by(userId=user_id)
        
        # Apply filters if provided
        if completed is not None:
            query = query.filter_by(completed=(completed.lower() == 'true'))
        if priority:
            query = query.filter_by(priority=priority)
        if category:
            query = query.filter_by(category=category)
        
        # Order by due date, with null values at the end
        tasks = query.order_by(Task.dueDate.asc().nullslast()).all()
        
        return jsonify([task.to_dict() for task in tasks])
    except Exception as e:
        return jsonify({'error': str(e) or 'Server error'}), 500

@tasks_bp.route('/', methods=['POST'])
@jwt_required()
def create_task():
    """Create a new task"""
    try:
        user_id = get_jwt_identity()
        data = request.json
        
        if not data or not data.get('title'):
            raise BadRequest('Title is required')
        
        # Create task
        task = Task(
            title=data['title'],
            userId=user_id,
            description=data.get('description'),
            priority=data.get('priority', 'medium'),
            category=data.get('category')
        )
        
        # Parse date fields
        if data.get('dueDate'):
            task.dueDate = datetime.fromisoformat(data['dueDate'].replace('Z', '+00:00'))
        if data.get('reminder'):
            task.reminder = datetime.fromisoformat(data['reminder'].replace('Z', '+00:00'))
        
        # Save to database
        db.session.add(task)
        db.session.commit()
        
        return jsonify(task.to_dict()), 201
    except BadRequest as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e) or 'Server error'}), 500

@tasks_bp.route('/<int:task_id>', methods=['GET'])
@jwt_required()
def get_task(task_id):
    """Get a specific task"""
    try:
        user_id = get_jwt_identity()
        task = Task.query.get(task_id)
        
        if not task:
            raise NotFound('Task not found')
        
        # Check ownership
        if task.userId != user_id:
            raise Forbidden('Not authorized to access this task')
        
        return jsonify(task.to_dict())
    except (NotFound, Forbidden) as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        return jsonify({'error': str(e) or 'Server error'}), 500

@tasks_bp.route('/<int:task_id>', methods=['PUT'])
@jwt_required()
def update_task(task_id):
    """Update a task"""
    try:
        user_id = get_jwt_identity()
        task = Task.query.get(task_id)
        
        if not task:
            raise NotFound('Task not found')
        
        # Check ownership
        if task.userId != user_id:
            raise Forbidden('Not authorized to update this task')
        
        data = request.json
        if not data:
            raise BadRequest('Request body is required')
        
        # Update fields
        if 'title' in data:
            task.title = data['title']
        if 'description' in data:
            task.description = data['description']
        if 'completed' in data:
            task.completed = data['completed']
        if 'priority' in data:
            task.priority = data['priority']
        if 'category' in data:
            task.category = data['category']
        if 'dueDate' in data:
            task.dueDate = datetime.fromisoformat(data['dueDate'].replace('Z', '+00:00')) if data['dueDate'] else None
        if 'reminder' in data:
            task.reminder = datetime.fromisoformat(data['reminder'].replace('Z', '+00:00')) if data['reminder'] else None
        
        db.session.commit()
        return jsonify(task.to_dict())
    except (BadRequest, NotFound, Forbidden) as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e) or 'Server error'}), 500

@tasks_bp.route('/<int:task_id>', methods=['DELETE'])
@jwt_required()
def delete_task(task_id):
    """Delete a task"""
    try:
        user_id = get_jwt_identity()
        task = Task.query.get(task_id)
        
        if not task:
            raise NotFound('Task not found')
        
        # Check ownership
        if task.userId != user_id:
            raise Forbidden('Not authorized to delete this task')
        
        db.session.delete(task)
        db.session.commit()
        
        return jsonify({'message': 'Task deleted successfully'})
    except (NotFound, Forbidden) as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e) or 'Server error'}), 500 